### PR TITLE
Port YouTube cog

### PR DESCRIPTION
**Pull request by parafoxia**
This PR ports the YouTube cog into V2.

Currently, the following commands have been implemented:

* `/youtube video link` -- allows users to search for a video using the title (and autocomplete) and return a link
* `/youtube video information` -- allows users to view more detailed information for a video

### CODE REVIEW

**Comment by Jonxslays** [comment]
Add a daily recurring task to check for new videos until a more robust implementation is in place.

File: carberretta/extensions/youtube.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+@plugin.listener(hikari.StartedEvent)
+async def on_started(_: hikari.StartedEvent) -> None:
+    # There is no way to get all videos directly from the API without
+    # either (1) going through OAuth, or (2) brute forcing, with an
+    # eventual cap on 500 results. Scraping isn't exactly ideal, but
+    # it only happens once on startup, and it works, so...get meme'd.
+
+    log.info("Searching for videos...")
+    videos = scrapetube.get_channel(Config.YOUTUBE_CHANNEL_ID)
```
Comment (parafoxia) [inline carberretta/extensions/youtube.py:88]: Somehow run all this stuff in an executor.

File: carberretta/extensions/youtube.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+@lightbulb.implements(lightbulb.SlashCommandGroup)
+async def cmd_youtube(_: lightbulb.SlashContext) -> None:
+    ...
+
+
+@cmd_youtube.child
+@lightbulb.option("title", "The title of the video.", autocomplete=True)
+@lightbulb.command("link", "Link a video (use if directing someone to watch a video).")
+@lightbulb.implements(lightbulb.SlashSubCommand)
+async def cmd_youtube_link(ctx: lightbulb.SlashContext) -> None:
```
Comment (parafoxia) [inline carberretta/extensions/youtube.py:106]: Add option to set the maximum number of results displayed (min = 5/10, max=25, default=10).

**Review by Jonxslays — DISMISSED** [review]
looks good

File: carberretta/bot.py [hunk 3]
```diff
@@ -63,8 +65,12 @@
 
 @bot.listen(hikari.StartingEvent)
 async def on_starting(_: hikari.StartingEvent) -> None:
+    cookie_jar = CookieJar(loop=asyncio.get_running_loop())
```
Comment (parafoxia) [inline carberretta/bot.py:68]: These cookies can be removed.

File: carberretta/extensions/youtube.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+from carberretta import Config
+from carberretta.utils import helpers
+
+if t.TYPE_CHECKING:
+    from aiohttp import ClientSession
+
+plugin = lightbulb.Plugin("YouTube", include_datastore=True)
+log = logging.getLogger(__name__)
+
+BROWSE_ENDPOINT = "https://www.youtube.com/youtubei/v1/browse"
```
Comment (Jonxslays) [inline carberretta/extensions/youtube.py:52]: I feel like this url should match the one below without the www

File: carberretta/extensions/youtube.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    "?part=brandingSettings%2Csnippet%2Cstatistics"
+)
+VIDEO_URL = (
+    "https://www.googleapis.com/youtube/v3/videos"
+    "?part=contentDetails%2Csnippet%2Cstatistics"
+)
+WATCH_URL = "https://youtube.com/watch?v="
+
+
+def _similarity(s1: str, s2: str, **_: t.Any) -> float:
```
Comment (Jonxslays) [inline carberretta/extensions/youtube.py:64]: Is there a particular reason you accept kwargs here but discard them?
I assume something to do with the args rapidfuzz passes but maybe worth a comment?

File: carberretta/extensions/youtube.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+            scorer=_similarity,
+            limit=10,
+            score_cutoff=0.5,
+        )
+    ]
+
+
+def _create_directory(kind: str) -> None:
+    url = (
+        f"https://www.youtube.com/channel/{Config.YOUTUBE_CHANNEL_ID}/"
```
Comment (Jonxslays) [inline carberretta/extensions/youtube.py:101]: Same as above with the www, maybe you should just extract the base URL into its own var.

File: carberretta/extensions/youtube.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    return _compile_options(opt.value, plugin.d.video_directory)
+
+
+@cmd_youtube_video.child
+@lightbulb.option(
+    "title",
+    "The title of the video you want to view information about.",
+    autocomplete=True,
+)
+@lightbulb.command("information", "View information about an video.")
```
Comment (Jonxslays) [inline carberretta/extensions/youtube.py:162]:
```suggestion
@lightbulb.command("information", "View information about a video.")
```

File: carberretta/extensions/youtube.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+            )
+            return
+
+        data = (await resp.json())["items"][0]
+
+    thumbnails: dict[str, dict[str, str]] = data["snippet"]["thumbnails"]
+    published = int(du_parse(data["snippet"]["publishedAt"]).timestamp())
+
+    def _fmt_duration(dur: str) -> str:
+        return dur.replace("H", "h ").replace("M", "' ").replace("S", '"')[2:]
```
Comment (Jonxslays) [inline carberretta/extensions/youtube.py:185]: lambda?

**Review by Jonxslays — CHANGES_REQUESTED** [review]
Mostly also nit picks, but 1 cmd description change.

File: carberretta/extensions/youtube.py (hunk 4 above)
Comment (parafoxia) [reply carberretta/extensions/youtube.py:52]: Yeah probably, good spot.

File: carberretta/extensions/youtube.py (hunk 5 above)
Comment (parafoxia) [reply carberretta/extensions/youtube.py:64]: Yeah it passes some kwargs we don't need so I just consume them all.

File: carberretta/extensions/youtube.py (hunk 6 above)
Comment (parafoxia) [reply carberretta/extensions/youtube.py:101]: I'm not sure since it never needs to be suffixed with anything.

Comment (Jonxslays) [reply carberretta/extensions/youtube.py:101]: It's used a few times throughout the code with different suffixes like `/youtubei/...`, `/channel/...` and `/watch?...`. in any case not a big deal.

File: carberretta/extensions/youtube.py (hunk 4 above)
Comment (parafoxia) [reply carberretta/extensions/youtube.py:52]: So the WWW here is necessary -- without it, it completely breaks. How odd. I've just added WWW to them all now lmao.

**Review by Jonxslays — APPROVED** [review]
`MINE_URL` is sussy

**Comment by parafoxia** [comment]
It's consistent with how all YouTube APIs typically denote "your something" in authorised requests (passing "MINE" as a parameter tells it to use your channel ID).

**Pull request merged**